weak_delegate: @WKExtensionDelegateAdaptor should be non-triggering

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

WKExtensionDelegateAdaptor triggers weak_delegate but should not.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
% swiftlint
Linting Swift files in current working directory
...
Linting 'Application.swift' (12/14)
...
/Users/username/Developer/Applications/ApplicationName/Sources/Application/Application.swift:19:57: warning: Weak Delegate Violation: Delegates should be weak to avoid reference cycles. (weak_delegate)
Done linting! Found 1 violation, 0 serious in 14 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.43.1
* Installation method used (Homebrew, CocoaPods, building from source, etc)? pkg install
* Paste your configuration file:

```yml
disabled_rules:
  - line_length

trailing_whitespace:
  ignores_empty_lines: true

```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
* Which Xcode version are you using (check `xcodebuild -version`)? Xcode 12.5 Build version 12E262
* Do you have a sample that shows the issue? See example below

```swift
import SwiftUI

@main
struct Application: App {

    #if os(iOS)

    @UIApplicationDelegateAdaptor(ApplicationDelegate.self) var applicationDelegate

    #elseif os(watchOS)

    @WKExtensionDelegateAdaptor(ExtensionDelegate.self) var extensionDelegate

    #endif

    let persistenceController = PersistenceController.shared

    @SceneBuilder var body: some Scene {
        WindowGroup {
            ContentView()
                .environment(\.managedObjectContext, persistenceController.container.viewContext)
        }

        #if os(watchOS)

        WKNotificationScene(controller: NotificationController.self, category: "myCategory")

        #endif
    }
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -590,6 +590,7 @@ macOS < 12.
 * Don't report `@NSApplicationDelegateAdaptor` and `@WKExtensionDelegateAdaptor` statements in `weak-delegate` rule.
   [Till Hainbach](https://github.com/tillhainbach)
   [#3598](https://github.com/realm/SwiftLint/issues/3456)
+  [#3611](https://github.com/realm/SwiftLint/issues/3611)
 
 #### Bug Fixes
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -67,6 +67,12 @@
   [Chris Brakebill](https://github.com/braker1nine)
   [#2708](https://github.com/realm/SwiftLint/issues/2708)
 
+* Do not report variables annotated with `@NSApplicationDelegateAdaptor` and
+  `@WKExtensionDelegateAdaptor` in `weak_delegate` rule.  
+  [Till Hainbach](https://github.com/tillhainbach)
+  [#3598](https://github.com/realm/SwiftLint/issues/3456)
+  [#3611](https://github.com/realm/SwiftLint/issues/3611)
+
 ## 0.49.1: Buanderie Principale
 
 _Note: The default branch for the SwiftLint git repository was renamed from
@@ -587,10 +593,7 @@ macOS < 12.
 
 #### Enhancements
 
-* Don't report `@NSApplicationDelegateAdaptor` and `@WKExtensionDelegateAdaptor` statements in `weak-delegate` rule.
-  [Till Hainbach](https://github.com/tillhainbach)
-  [#3598](https://github.com/realm/SwiftLint/issues/3456)
-  [#3611](https://github.com/realm/SwiftLint/issues/3611)
+* None.
 
 #### Bug Fixes
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -587,7 +587,7 @@ macOS < 12.
 
 #### Enhancements
 
-* Don't report `@NSApplicationDelegateAdaptor` statements in
+* Don't report `@NSApplicationDelegateAdaptor` and `@WKExtensionDelegateAdaptor` statements in `weak-delegate` rule.
   [Till Hainbach](https://github.com/tillhainbach)
   [#3598](https://github.com/realm/SwiftLint/issues/3456)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -587,7 +587,9 @@ macOS < 12.
 
 #### Enhancements
 
-* None.
+* Don't report `@NSApplicationDelegateAdaptor` statements in
+  [Till Hainbach](https://github.com/tillhainbach)
+  [#3598](https://github.com/realm/SwiftLint/issues/3456)
 
 #### Bug Fixes
 

--- a/Source/SwiftLintFramework/Extensions/String+SwiftLint.swift
+++ b/Source/SwiftLintFramework/Extensions/String+SwiftLint.swift
@@ -122,4 +122,8 @@ extension String {
             }
         }
     }
+
+    internal func hasAnyPrefix(of prefixes: [String]) -> Bool {
+        prefixes.contains(where: hasPrefix)
+    }
 }

--- a/Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift
@@ -26,7 +26,8 @@ public struct WeakDelegateRule: OptInRule, ASTRule, ConfigurationProviderRule {
             Example("class Foo {\n protocol P {\n var delegate: AnyObject? { get set }\n}\n}\n"),
             Example("class Foo {\n var computedDelegate: ComputedDelegate {\n return bar() \n} \n}"),
             Example("struct Foo {\n @UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate \n}"),
-            Example("struct Foo {\n @NSApplicationDelegateAdaptor(AppDelegate.self) var appDelegate \n}")
+            Example("struct Foo {\n @NSApplicationDelegateAdaptor(AppDelegate.self) var appDelegate \n}"),
+            Example("struct Foo {\n @WKExtensionDelegateAdaptor(ExtensionDelegate.self) var extensionDelegate \n}")
         ],
         triggeringExamples: [
             Example("class Foo {\n  ↓var delegate: SomeProtocol?\n}\n"),
@@ -75,7 +76,9 @@ public struct WeakDelegateRule: OptInRule, ASTRule, ConfigurationProviderRule {
                 let offset = attribute.offset,
                 let length = attribute.length,
                 let value = file.stringView.substringWithByteRange(ByteRange(location: offset, length: length)),
-                (value.hasPrefix("@UIApplicationDelegateAdaptor") || value.hasPrefix("@NSApplicationDelegateAdaptor")) {
+                (value.hasPrefix("@UIApplicationDelegateAdaptor")
+                  || value.hasPrefix("@NSApplicationDelegateAdaptor")
+                  || value.hasPrefix("@WKExtensionDelegateAdaptor")) {
                 return []
             }
         }

--- a/Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift
@@ -35,6 +35,12 @@ public struct WeakDelegateRule: OptInRule, ASTRule, ConfigurationProviderRule {
         ]
     )
 
+    private static let ignoredAnnotations = [
+        "@UIApplicationDelegateAdaptor",
+        "@NSApplicationDelegateAdaptor",
+        "@WKExtensionDelegateAdaptor"
+    ]
+
     public func validate(file: SwiftLintFile, kind: SwiftDeclarationKind,
                          dictionary: SourceKittenDictionary) -> [StyleViolation] {
         return violationRanges(in: file, kind: kind, dictionary: dictionary).map {
@@ -69,16 +75,13 @@ public struct WeakDelegateRule: OptInRule, ASTRule, ConfigurationProviderRule {
         // Check if non-computed
         let isComputed = (dictionary.bodyLength ?? 0) > 0
         guard !isComputed else { return [] }
-
-        // Check for UIApplicationDelegateAdaptor
+        // Check for annotations
         for attribute in dictionary.swiftAttributes {
             if
                 let offset = attribute.offset,
                 let length = attribute.length,
                 let value = file.stringView.substringWithByteRange(ByteRange(location: offset, length: length)),
-                (value.hasPrefix("@UIApplicationDelegateAdaptor")
-                  || value.hasPrefix("@NSApplicationDelegateAdaptor")
-                  || value.hasPrefix("@WKExtensionDelegateAdaptor")) {
+                value.hasAnyPrefix(of: Self.ignoredAnnotations) {
                 return []
             }
         }

--- a/Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift
+++ b/Source/SwiftLintFramework/Rules/Lint/WeakDelegateRule.swift
@@ -25,7 +25,8 @@ public struct WeakDelegateRule: OptInRule, ASTRule, ConfigurationProviderRule {
             Example("protocol P {\n var delegate: AnyObject? { get set }\n}\n"),
             Example("class Foo {\n protocol P {\n var delegate: AnyObject? { get set }\n}\n}\n"),
             Example("class Foo {\n var computedDelegate: ComputedDelegate {\n return bar() \n} \n}"),
-            Example("struct Foo {\n @UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate \n}")
+            Example("struct Foo {\n @UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate \n}"),
+            Example("struct Foo {\n @NSApplicationDelegateAdaptor(AppDelegate.self) var appDelegate \n}")
         ],
         triggeringExamples: [
             Example("class Foo {\n  ↓var delegate: SomeProtocol?\n}\n"),
@@ -74,7 +75,7 @@ public struct WeakDelegateRule: OptInRule, ASTRule, ConfigurationProviderRule {
                 let offset = attribute.offset,
                 let length = attribute.length,
                 let value = file.stringView.substringWithByteRange(ByteRange(location: offset, length: length)),
-                value.hasPrefix("@UIApplicationDelegateAdaptor") {
+                (value.hasPrefix("@UIApplicationDelegateAdaptor") || value.hasPrefix("@NSApplicationDelegateAdaptor")) {
                 return []
             }
         }

